### PR TITLE
Add action graph API

### PR DESCRIPTION
**Pull request by jacobperron**
As a prerequisite, I've  moved the names and types functions to `rclpy_common` for reuse in the rclpy action C extension.

Connects to ros2/ros2cli#202

### CODE REVIEW

**Comment by jacobperron** [comment]
Testing `rclpy`:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6590/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2928/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5409/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6413/)

File: rclpy/src/rclpy/_rclpy.c [hunk 1]
```diff
@@ -3375,18 +3306,8 @@ rclpy_get_subscriber_names_and_types_by_node(PyObject * Py_UNUSED(self), PyObjec
...
-  }
-
-  if (!__convert_names_and_types(topic_names_and_types, pytopic_names_and_types)) {
-    __cleanup_names_and_types(&topic_names_and_types);
-    Py_DECREF(pytopic_names_and_types);
-    return NULL;
-  }
-  __cleanup_names_and_types(&topic_names_and_types);
+  PyObject * pytopic_names_and_types = rclpy_convert_to_py_names_and_types(&topic_names_and_types);
+  rclpy_names_and_types_fini(&topic_names_and_types);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rclpy/src/rclpy/_rclpy.c]: Good catch. Fixed 3886c34

File: rclpy/src/rclpy/_rclpy_action.c [hunk 2]
```diff
@@ -1682,6 +1682,111 @@ rclpy_action_expire_goals(PyObject * Py_UNUSED(self), PyObject * args)
...
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(
+      PyExc_RuntimeError,
+      "Failed to get action client names and type: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  PyObject * pynames_and_types = rclpy_convert_to_py_names_and_types(&names_and_types);
+  rclpy_names_and_types_fini(&names_and_types);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rclpy/src/rclpy/_rclpy_action.c]: 3886c34

File: rclpy/test/action/test_graph.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        get_names_and_types_func,
+        *args,
+        expected_num_names,
+        timeout=5.0
+    ):
+        # Since it can take some time for the ROS graph to update
+        # keep trying to get names and types until we get the desired number or timeout
+        start = time.monotonic()
+        end = start
+        while (end - start) < timeout:
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rclpy/test/action/test_graph.py]: Good catch. Fixed efa498d

File: rclpy/src/rclpy_common/src/common.c.orig [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/error_handling.h"
+
+#include "rclpy_common/common.h"
+
+<<<<<<< HEAD
```
Comment (sloretz) [inline rclpy/src/rclpy_common/src/common.c.orig]: Leftover from a merge conflict?

Comment (jacobperron) [reply rclpy/src/rclpy_common/src/common.c.orig]: Yeah, oops. Removed 8883a9e

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6670/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2989/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5478/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6504/)

File: rclpy/rclpy/action/__init__.py [hunk 5]
```diff
@@ -13,4 +13,7 @@
 # limitations under the License.
 
 from .client import ActionClient  # noqa
+from .graph import get_action_client_names_and_types_by_node  # noqa
+from .graph import get_action_names_and_types  # noqa
+from .graph import get_action_server_names_and_types_by_node  # noqa
```
Comment (dirk-thomas) [inline rclpy/rclpy/action/__init__.py]: Don't suppress all kinds of warnings but only one specific error code.

Comment (jacobperron) [reply rclpy/rclpy/action/__init__.py]: Done bb44498

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6681/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3000/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5489/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6511/)

**Pull request merged**